### PR TITLE
Arbitrary 32-byte contents in memos

### LINKED ISSUE #2660: Change memo field semantics
Currently the memo field is only a SHA-2. We'd like to be able to stash arbitrary info here. We need an RFC to figure out what we want to do here.

### LINKED ISSUE #2701: 32byte memo don't sha, just pass raw bytes
Short term we should be able to interpret these 32bytes as a string.

### PR DESCRIPTION
**Pull request by psteckler**
Allow arbitrary 32-byte strings in user command memos. Can still pass an arbitrary string (up to a max length) to digest.

Can use a `bytes` or `string` to create a memo, the length is checked. 

Length violations raise exceptions.

Perhaps problematic: some code from `Random_oracle.Digest` can't be used directly, so there's a modicum of copied code. There could be a functor to factor out that small amount of commonality, but I'm hesistant to introduce more functors.

Update: the code is different enough that it can't really be shared by functorization.

Closes #2660
Closes #2701

### CODE REVIEW

File: src/lib/coda_base/user_command_memo.ml [hunk 1]
```diff
@@ -1,21 +1,99 @@
...
+  if Int.(String.length s > max_digestible_string_length) then
+    raise Too_long_digestible_string ;
+  (Random_oracle.digest_string s :> t)
+
+let create_from_bytes32_exn bytes : t =
+  if not (Int.equal (Bytes.length bytes) length_in_bytes) then
+    raise Invalid_user_memo_length ;
+  Bytes.to_string bytes
+
+let create_from_string32_exn s : t =
```
Comment (bkase) [inline src/lib/coda_base/user_command_memo.ml]: Can you make versions of these that are safe as well? I think the graphql stuff should use the non exn versions too.

**Comment by cmr** [comment]
Hm. I feel like we should length-prefix these strings. Use the first byte to count the number of bytes in the message. The alternative right now is padding the string with some sentinel byte (probably 0/nul).

**Comment by psteckler** [comment]
@cmr Do you mean:
- for digests, having a 33-byte string, with the 0th byte equal to 32
- for strings, allowing strings shorter than 32, prepend a byte equal to their length

**Comment by cmr** [comment]
Yeah that's what meant. But you writing it out made me realize it's not possible to distinguish a digest from a string under this design. Maybe we should add two bytes then, one for type and one for length. I propose:

extend field to 34 bytes. first byte is a type.

types: 0 for blake2 digest, 1 for string. reject all other values in transaction validation.

for type 0: write out the 32 digest bytes to the remainder of the field. pad it out by setting the final byte to zero.
for type 1: first non-version byte contains a length. write the string contents in the remainder of the field, pad remainder with zeros.

For both types, reject the transaction if the memo is incorrectly padded.

I think when it comes to transporting these values in/out of the SNARK, we need a fixed, uniform length.

**Comment by psteckler** [comment]
Made changes requested by @cmr and @bkase.

Will add some tests.

**Comment by cmr** [comment]
In chat, @enolan suggested keeping the length byte for type 0 and hardcoding it to 32. I don't disagree, and that's what's implemented now.

File: src/lib/coda_base/user_command_memo.ml [hunk 2]
```diff
@@ -1,21 +1,168 @@
...
+    Fold_lib.Fold.(to_list (group3 ~default:Boolean.false_ (of_array t)))
+
+  let constant unchecked =
+    assert (Int.(String.length (unchecked :> string) = digest_length + 2)) ;
+    Array.map
+      (Blake2.string_to_bits (unchecked :> string))
+      ~f:Boolean.var_of_value
+end
+
+let length_in_bits = 8 * digest_length
```
Comment (cmr) [inline src/lib/coda_base/user_command_memo.ml]: `8 * (digest_length+2)`, right?

File: src/lib/coda_base/user_command_memo.ml [hunk 3]
```diff
@@ -1,21 +1,168 @@
...
+
+let length_in_triples = (length_in_bits + 2) / 3
```
Comment (cmr) [inline src/lib/coda_base/user_command_memo.ml]: I see this `+2` is also in the `random_oracle_base.ml`, I guess to make sure the length in triples gets rounded up and not down by the integer division?

File: src/lib/coda_base/user_command_memo.ml [hunk 4]
```diff
@@ -1,21 +1,168 @@
...
+            let b = (Char.to_int t.[i / 8] lsr (i mod 8)) land 1 = 1 in
+            go (f acc b) (i + 1)
+        in
+        go init 0 ) }
+
+let fold t = Fold_lib.Fold.group3 ~default:false (fold_bits t)
+
+let typ : (Checked.t, t) Typ.t =
+  Typ.transport
+    (Typ.array ~length:Blake2.digest_size_in_bits Boolean.typ)
```
Comment (cmr) [inline src/lib/coda_base/user_command_memo.ml]: This should be `length_in_bits`

File: src/lib/coda_base/user_command_payload.ml [hunk 5]
```diff
@@ -67,8 +67,9 @@ module Common = struct
     let%map fee = Currency.Fee.gen
     and nonce = Account_nonce.gen
     and memo =
-      String.gen_with_length Memo.max_size_in_bytes Char.quickcheck_generator
-      >>| Memo.create_exn
+      String.gen_with_length Memo.max_digestible_string_length
```
Comment (cmr) [inline src/lib/coda_base/user_command_payload.ml]: Can you quickcheck a bool and use that to determine whether to use a bytes memo or a digest memo?

File: src/lib/coda_base/user_command_memo.ml (hunk 3 above)
Comment (psteckler) [reply src/lib/coda_base/user_command_memo.ml]: I believe so.

File: src/lib/coda_base/user_command_payload.ml (hunk 5 above)
Comment (psteckler) [reply src/lib/coda_base/user_command_payload.ml]: Can do.

**Comment by bkase** [comment]
> Maybe we should add two bytes then, one for type and one for length

I think being able to distinguish between the sha256 and readable string is nice, but we can also live without it and just print it as a sha hash if there are any non-unicode bytes. For 32 characters in a row, the probability that all the bytes are accidentally printable unicode bytes is less than 0.1% (and probably significantly so, but I don't want to expand this equation manually, and they only show the table up to 10 bytes) https://math.stackexchange.com/questions/750551/probability-that-random-byte-array-is-a-valid-utf-8-string

I suggested the approach of just changing what we store in those 32bytes to avoid making the larger change, as this is the sort of thing we'd prefer to have an RFC for I think

On the UI-side, it's pretty easy definitely better to have the prefix bytes, but this does impact SNARK performance (our state is ~1kbytes so it's less than a 1% impact).

Given that this is already implemented, I think we should tentatively go with it, but also create an RFC, and we can revert this change if the RFC doesn't pass. Given that you've implemented it @psteckler , would you like to make the RFC? If you're too busy, I'd be happy to work on it.

**Comment by bkase** [comment]
Yeah it's definitely hacky to just look and see if it's valid unicode

**Comment by psteckler** [comment]
> I think being able to distinguish between the sha256 and readable string is nice

Of course, the arbitrary bytes aren't necessarily human-readable.

**Comment by cmr** [comment]
If we don’t enforce any hard requirements on memo validity, then this shouldn’t add many additional constraints, right? We can measure...

**Comment by psteckler** [comment]
In the RFC, I mentioned the possibility of distinguish arbitrary bytes from human-readable bytes, which we could add here, if it's a good idea.

In the tests, there a function `data` to pull out the memo data. Should that be exposed (could be useful for GraphQL)?

**Pull request merged**